test/Microsoft.NET.Restore.Tests/GivenThatWeWantToUseFrameworkRoslyn.cs

```diff
@@ -128,5 +128,37 @@ public void It_throws_an_error_when_the_package_is_not_downloaded()
             buildCommand.ExecuteWithoutRestore("/p:BuildWithNetFrameworkHostedCompiler=true")
                 .Should().Fail().And.HaveStdOutContaining("NETSDK1216");
         }
+
+        [FullMSBuildOnlyFact]
+        public void It_throws_a_warning_when_NuGetPackageRoot_is_empty()
+        {
+            const string testProjectName = "NetCoreApp";
+            var project = new TestProject
+            {
+                Name = testProjectName,
+                TargetFrameworks = "net6.0",
+            };
+
+            // simulate mismatched MSBuild versions
+            project.AdditionalProperties.Add("_IsDisjointMSBuildVersion", "true");
+
+            var testAsset = _testAssetsManager
+                .CreateTestProject(project);
+
+            NuGetConfigWriter.Write(testAsset.Path, TestContext.Current.TestPackages);
+
+            var customPackagesDir = Path.Combine(testAsset.Path, "nuget-packages");
+
+            var command = (MSBuildCommand)new MSBuildCommand(testAsset, "Restore;Build")
+                .WithEnvironmentVariable("NUGET_PACKAGES", customPackagesDir);
+            command.ExecuteWithoutRestore()
+                .Should().Pass().And.HaveStdOutContaining("NETSDK1221");
+
+            // The package is downloaded, but the targets cannot find the path to it
+            // because NuGetPackageRoot is empty during `/t:Restore;Build`.
+            // See https://github.com/dotnet/sdk/issues/43016.
+            var toolsetPackageDir = Path.Combine(customPackagesDir, "microsoft.net.sdk.compilers.toolset");
+            Assert.True(Directory.Exists(toolsetPackageDir));
```

Comment:
Nit: you'll get better information about why the test failed in the logs if you do this instead:
```suggestion
            new DirectoryInfo(toolsetPackageDir).Should().Exist();
```

Reply:
Thanks, I'm not used to fluent assertions being available in repos I work in.